File: frontend/src/api/session.ts
import { readEventStream, request, sendFile } from "./http";

// 输入框里暂存的附件。text 只在发送时交给模型，气泡不展示。

export type ActivityTodo = {
  id: string;
  label: string;
  status?: "pending" | "active" | "complete" | "cancelled";
};

export type ClarificationQuestion = {
  id: string;
  prompt: string;
  options: { id: string; label: string }[];
};

export type ClarificationAnswer = {
  id: string;
  option_id: string;
  label: string;
  prompt: string;
};

export type ChatAttachment = {
  name: string;
  size: number;
  text?: string;
};

export type ChatMessage = {
  id: string;
  role: "user" | "assistant" | "system";
  content: string;
  extra?: {
    kind?: string;
    question_set_id?: string;
    questions?: { stem?: string; ordinal?: number; id?: string; prompt?: string; options?: { id: string; label: string }[] }[];
    actions?: string[];
    todos?: ActivityTodo[];
    answers?: ClarificationAnswer[];
    files?: { name: string; size: number }[];
    thinking?: string;
    reasoning?: string;
  } | null;
  created_at: string;
};

export type ChatStreamEvent =
  | { type: "blocked" }
  | { type: "meta"; session_id: string; extra?: ChatMessage["extra"] }
  | { type: "thinking"; text: string }
  | { type: "reasoning"; text: string }
  | { type: "delta"; text: string }
  | { type: "done"; session: ChatSession };

export type ChatSession = {
  id: string;
  title: string;
  job_title?: string | null;
  created_at: string;
  updated_at: string;
  messages?: ChatMessage[];
};

export const sessionApi = {
  sessions: () => request<ChatSession[]>("/api/sessions"),
  session: (id: string) => request<ChatSession>(`/api/sessions/${id}`),
  createSession: () => request<ChatSession>("/api/sessions", { method: "POST" }),
  clearSession: (id: string) => request<ChatSession>(`/api/sessions/${id}/clear`, { method: "POST" }),
  deleteSession: (id: string) => request<{ ok: string }>(`/api/sessions/${id}`, { method: "DELETE" }),
  chat: (
    content: string,
    sessionId?: string,
    answers?: ClarificationAnswer[],
    signal?: AbortSignal,
    attachments?: ChatAttachment[],
  ) =>
    request<ChatSession>("/api/chat", {
      method: "POST",
      body: JSON.stringify({ content, session_id: sessionId, answers: answers ?? [], attachments: attachments ?? [] }),
      signal,
    }),
  prepareChatFile: (file: File) => sendFile<ChatAttachment>("/api/chat/prepare", file),
  streamChat: (
    content: string,
    sessionId: string | undefined,
    answers: ClarificationAnswer[] | undefined,
    attachments: ChatAttachment[] | undefined,
    onEvent: (event: ChatStreamEvent) => void,
    signal?: AbortSignal,
  ) =>
    readEventStream(
      "/api/chat/stream",
      { content, session_id: sessionId, answers: answers ?? [], attachments: attachments ?? [] },
      (event) => onEvent(event as ChatStreamEvent),
      signal,
    ),
};

File: backend/app/schemas/session.py
"""Request and response shapes for the session cockpit."""

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class ChatSessionOut(BaseModel):
    id: str
    title: str
    job_title: str | None = None
    created_at: datetime
    updated_at: datetime


class ChatMessageOut(BaseModel):
    id: str
    role: Literal["user", "assistant", "system"]
    content: str
    extra: dict[str, Any] | None = None
    created_at: datetime


class ChatSessionDetail(ChatSessionOut):
    messages: list[ChatMessageOut] = Field(default_factory=list)


class ClarificationAnswerIn(BaseModel):
    id: str
    option_id: str
    label: str = ""
    prompt: str = ""


class ChatAttachmentIn(BaseModel):
    """随这条消息一起提交的附件。正文已在前端读出，气泡只展示文件名。"""

    name: str
    size: int = 0
    text: str = ""


class ChatSendIn(BaseModel):
    content: str = ""
    session_id: str | None = None
    # 仅用于回复上一轮澄清题。普通消息留空。
    answers: list[ClarificationAnswerIn] = Field(default_factory=list)
    attachments: list[ChatAttachmentIn] = Field(default_factory=list)
